[KOGITO-9940] Add documentation for Jobs Service support in the Operator (#530)
apache/incubator-kie-kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/operator/enabling-jobs-service.adoc

= Managing Jobs Service with the Operator
:compat-mode!:
// Metadata:
:description: Configure Jobs Service using the `SonataFlowPlatform` CR.
:keywords: sonataflow, serverless, operator, kubernetes, jobs service

This document describes how to configure the Jobs Service instance usimg the SonataFlowPlarform CR.

> == Automate the Jobs Service instance management with the `SonataFlow` Operator

Reviewer: ```suggestion
== Automate the Jobs service instance management with the `SonataFlow` Operator
```

> While it is possible to deploy the Jobs Service manually, the Operator also provides a convenient way to combine it with the

Reviewer: ```suggestion
It is possible to deploy the Jobs Service manually, but the Operator also provides a convenient way to combine it with the
```

namespace configuration via the `SonataFlowPlatform` CR. With this approach, the Operator will take care of configuring the
Jobs Service instance and ensure it is in sync with the specification in the CR. When the operator manages the Jobs Service lifecycle,
it will inject properties in `SonataFlow` workflows at creation time to enable the workflows to communicate with the Jobs Service

> instance during their execution, removing the need to add these properties as part of the `SonataFlow` workflow CR instance.

Reviewer: We can rewrite the above paragraph like this: 
```suggestion
It is possible to deploy the Jobs Service manually, but the Operator also provides a convenient way to combine it with the namespace configuration via the `SonataFlowPlatform` CR. This approach allows the Operator to configure the Jobs Service and make sure it is in line with the Specification in CR. When the Operator manages the jobs service lifecycle, it injects properties at creation time in the `SonataFlow` workflows to communicate with the Jobs Service during execution, eliminating the need to include these properties in the workflow CR instance of the `SonataFlow` workflow.
```

> It is possible to deploy the Jobs Service manually, but the Operator also provides a convenient way to combine it with the namespace configuration via the `SonataFlowPlatform` CR. This approach allows the Operator to configure the Jobs Service and make sure it is in line with the Specification in CR. When the Operator manages the jobs service lifecycle, it injects properties at creation time in the `SonataFlow` workflows to communicate with the Jobs Service during execution, eliminating the need to include these properties in the workflow CR instance of the `SonataFlow` workflow.

Reviewer 2: @kaldesai not sure if you saw this, can you please take a look?

> It is possible to deploy the Jobs Service manually, but the Operator also provides a convenient way to combine it with the namespace configuration via the `SonataFlowPlatform` CR. This approach allows the Operator to configure the Jobs Service and make sure it is in line with the Specification in CR. When the Operator manages the jobs service lifecycle, it injects properties at creation time in the `SonataFlow` workflows to communicate with the Jobs Service during execution, eliminating the need to include these properties in the workflow CR instance of the `SonataFlow` workflow.

Reviewer: @ricardozanini, How about now?
I tried to minimize the content without changing the meaning of the paragraph. Let me know WDYT :) Otherwise, we can change it again. 
 
`It is possible to deploy the Jobs Service manually, leveraging the Operator offers a more seamless integration by combining it with namespace configuration through the SonataFlowPlatform Custom Resource (CR). When the Operator oversees the lifecycle of the jobs service, it automatically injects necessary properties during creation into the SonataFlow workflows. This integration eliminates the need for including these properties within the SonataFlow workflow CR instance, simplifying workflow management.`

> It is possible to deploy the Jobs Service manually, but the Operator also provides a convenient way to combine it with the namespace configuration via the `SonataFlowPlatform` CR. This approach allows the Operator to configure the Jobs Service and make sure it is in line with the Specification in CR. When the Operator manages the jobs service lifecycle, it injects properties at creation time in the `SonataFlow` workflows to communicate with the Jobs Service during execution, eliminating the need to include these properties in the workflow CR instance of the `SonataFlow` workflow.

Author: Done

== Configuring Jobs Service in the SonataFlowPlatformCR

To enable the deployment of a Jobs Service instance, the `SonataFlowPlatform` CRD exposes a set of fields that allow the user to 
configure the running instance. 

==== Ephemeral persistence
The basic runtime is to deploy the Jobs Service with an ephemeral backend running in the same container
as the Jobs Service runtime.

[source,yaml,subs="attributes+"]
---
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlowPlatform
metadata:
  name: sonataflow-platform
spec:
  services:
    jobService: {}
---

When executing this manifest, the operator will reconcile generating a pod hosting the Jobs Service:

[source,shell,subs="attributes+"]
---
$>kubectl get pod -n sonataflow
NAME                                               READY   STATUS    RESTARTS   AGE
sonataflow-platform-jobs-service-cdf85d969-sbwkj   1/1     Running   0          108s
---

> Keep in mind that this setup is not recommended for production environments, chiefly because the data is not persisted when the pod restarts.

Reviewer: ```suggestion
Keep in mind that this setup is not recommended for production environments, especially because the data does not persist when the pod restarts.
```

==== Using an existing postgreSQL service

> For robust environments it is recommened to use an dedicated database service and configure Jobs Service to make use of it. Currently, the Jobs Service

Reviewer: ```suggestion
For robust environments, it is recommended to use a dedicated database service and configure Jobs Service to make use of it. Currently, the Jobs Service
```

only supports PostgreSQL database.

> Configuring Jobs Service to communicate with an existing PostgreSQL instance is supported in two ways. In both cases it requires providing the persistence

Reviewer: ```suggestion
Configuring Jobs Service to communicate with an existing PostgreSQL instance is supported in two ways. In both the cases it requires providing the persistence
```

Reviewer 3: Not sure if this sounds right, should the suggestion be `In both of the cases`?

Author: According to [this](https://forum.thefreedictionary.com/postst216720_Both--Both-the.aspx) website, the usage of "both the <noun>" is relevant when the `noun` is singular:

> Both the sentences are grammatically correct. "The" is usually not necessary when the ensuing noun is plural and only one noun is used e.g. both sisters, both pictures and both children.
> 
> "The" is required in the following pattern in singular: Both the elephant and the bison, both the first and the last book, both the boy and the girl...
> 
> Additionally, "the" is required in the following pattern: Both the sisters and the brothers. You do not say: Both sisters and the brothers.

Disclaimer: I'm not a native speaker

configuration, one by using the Jobs Service's persistence field and the other one using the persistence field defined in the `SonataFlowPlatform` CR
deployed in the same namespace.

> By default, the persistence specification defined in the `SonataFlow` workflow's CR takes precedence over the one in the `SonataFlowPlatform` persistence's specification.

Reviewer: ```suggestion
By default, the persistence specification defined in the `SonataFlow` workflow's CR takes priority over the one in the `SonataFlowPlatform` persistence specification.
```

> ===== Using the persistence field defined in the `SonataFlowPlatform` CR

Author: Depends on how you want to structure the documentation. Here I am taking the assumption that this document contains all relevant information on how configure the job service, including persistence. If you prefer to contain persistence in its own scope, then I would suggest to move it not in the workflow-persistence.adoc that you mention, but in a `platform-persistence.adoc` or `persistence.adoc` where it contains how to configure persistence for workflows and platform services.

Let me know which way you prefer.

> ===== Using the persistence field defined in the `SonataFlowPlatform` CR

Reviewer 2: Summon @domhanak since he's restructuring a few things here. I think we can move to platform-persistence.adoc, would make more sense. Thank you for the suggestions!

Using the persistence configuration in the `SonataFlowPlatform` CR located in the same namespace requires to have the `SonataFlow` CR persistence field configured
to have an empty `{}` value, signaling the Operator to derive the persistence from the active `SonataFlowPlatform`, when available. If no persistence is defined
the operator will fallback to the ephemeral persistence previously described.

[source,yaml,subs="attributes+"]
---
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlowPlatform
metadata:
  name: sonataflow-platform
spec:
  persistence:
    postgresql:
      secretRef:
        name: postgres-secrets
        userKey: POSTGRES_USER
        passwordKey: POSTGRES_PASSWORD
      serviceRef:
        name: postgres
        port: 5432
        databaseName: sonataflow
---

And the `SonataFlow` CR looks like this:

[source,yaml,subs="attributes+"]
---
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlow
metadata:
  name: callbackstatetimeouts
  annotations:
    sonataflow.org/description: Callback State Timeouts Example k8s
    sonataflow.org/version: 0.0.1
spec:
  persistence: {}
...
---

> When using the `jdbcUrl` field instead of `serviceRef`, the user is responsible of providing the correct JDBC URL connection value that does not contain a `database schema`

Reviewer: ```suggestion
When using the `jdbcUrl` field instead of `serviceRef`, the user is responsible for providing the correct JDBC URL connection value that does not contain a `database schema`
```

because the operator will use verbatim the contents of this field as the JDBC connection in the Jobs Service pod, and if it provides a schema that has been used or formatted 
by a different client, the pod will fail to run.

===== Using the persistency field inside the service specification

> You can specify define the persistence configuration directly in the Jobs Service specification. The structure is the same as in the `SonataFlowPlatform` CR and also

Reviewer: ```suggestion
You can specify and define the persistence configuration directly in the Jobs Service specification. The structure is the same as in the `SonataFlowPlatform` CR and also
```

> consist on the credentials to access the PostgreSQL instance, and the kubernetes service reference to generate the connectivity.

Reviewer: ```suggestion
consist of the credentials to access the PostgreSQL instance, and the kubernetes service reference to generate the connectivity.
```

> ===== Using the persistency field inside the service specification

Reviewer 2: ```suggestion
===== Using the persistence field inside the service specification
```

[source,yaml,subs="attributes+"]
---
apiVersion: sonataflow.org/v1alpha08

> kind: SonataFlow

Reviewer 2: This should be a `SonataFlowPlatform`, no?

and referencing the database `sonataflow` and schema `jobs-service`. By default, when the Jobs Service pod starts it will recreate the schema in the given location
if it doesn't exist.

The values of `POSTGRES_USER` and `POSTGRES_PASSWORD` point to the keys in the secret that contains the PostgreSQL credentials.

[source,shell,subs="attributes+"]
---
$>kubectl describe secrets postgres-secrets
Name:         postgres-secrets
Namespace:    default
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
PGDATA:             28 bytes
POSTGRES_DATABASE:  10 bytes
POSTGRES_PASSWORD:  10 bytes
POSTGRES_USER:      10 bytes
---

Putting everything together will render an output similar to this one below:

[source,shell,subs="attributes+"]
---
$>kubectl get pod -w -n sonataflow
NAME                                                READY   STATUS    RESTARTS   AGE
postgres-64c9ddb6bc-v96qb                           1/1     Running   0          14m
sonataflow-platform-jobs-service-79bb6bc67f-gvcg4   1/1     Running   0           1m
---

== Conclusion

> The Operator extends its scope to manage the lifecycle of the Jobs Service instance, thus removing the burden to the users and allowing them to focus on the

Reviewer: ```suggestion
The Operator extends its scope to manage the lifecycle of the Jobs Service instance, thus removing the burden on the users and allowing them to focus on the
```

> implementation of the `Sonataworkflows. It takes care of managing the Jobs Service deployment, dynamically configure its application properties

Reviewer: ```suggestion
implementation of the `Sonataworkflows. It takes care of managing the Jobs Service deployment, dynamically configures its application properties
```
